source/Ship.cpp

```diff
@@ -3710,9 +3710,25 @@ void Ship::DoGeneration()
 
 	isDisabled = isOverheated || hull < MinimumHull() || (!crew && RequiredCrew());
 
+	double coolingEfficiency = CoolingEfficiency();
+	heat -= coolingEfficiency * attributes.Get("cooling");
 	// Update ship supply levels.
 	if(isDisabled)
+	{
 		PauseAnimation();
+		// If overheated, but otherwise not disabled, apply active cooling.
+		if(!(hull < MinimumHull() || (!crew && RequiredCrew())))
+		{
+			// If overheated, heat must be >= 100%
+			// So there is no need to check it again when handling active cooling.
+			// Still need to check for energy though.
+			if(energy >= 0.)
+			{
+				heat -= coolingEfficiency * attributes.Get("active cooling");
```

Comment:
You seem to be doing a much simpler calculation than the active cooling check and calculations below, reason?

Reply:
This check is only being called when the ship is overheated. If you're overheated, you already know that your heat must be at or above 100%, so there's no need to include the heat-scaling portion of active cooling.

Reply:
What I mean is the spent energy stuff, with this calculation you have here a ship can get negative energy

Reply:
Energy is clamped to be positive immediately after this with `energy = max(0., energy);` so negative energy isn't an issue. Now I look at it, this does cause one minor difference which is that you get the full cooling on the single frame where you run out of energy, rather than only a proportion of it based on how much energy you had left. I'll fix that.

Reply:
Thanks